python_modules/dagster/dagster/_config/pythonic_config/__init__.py

```diff
@@ -115,7 +116,24 @@ def __setattr__(self, name: str, value: Any):
             object.__setattr__(self, name, value)
             return
 
-        return super().__setattr__(name, value)
+        try:
+            return super().__setattr__(name, value)
+        except TypeError as e:
+            if "is immutable and does not support item assignment" in str(e):
+                clsname = self.__class__.__name__
+                if isinstance(self, ConfigurableResourceFactory):
+                    raise DagsterInvalidInvocationError(
+                        f"'{clsname}' is a Pythonic resource and does not support item assignment."
+                        " If trying to set state on this resource, consider building a separate,"
+                        " stateful client class."
```

Comment:
I would mention:

1) I would mention how this relates to Pydantic.(e.g. this inherits from BaseModel with frozen=true)
2) I would be a little more prescriptive. Explain how the pattern is have a method on the ConfigurableResource that constructs the other object. 

Also consider catching ValueError here to detect the assignment of undeclared attributes:

```
class TestModel(Config):
    def set_something(self):
        self._something = "something"


model = TestModel()
model.set_something()
```

```ValueError: "TestModel" object has no field "_something"```

Reply:
updated error message + added valueerror catching